sdk/identity/Azure.Identity/src/ProcessRunner.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Identity
+{
+#pragma warning disable CA1001 // Types that own disposable fields should be disposable. Disposing of _process and _ctRegistration / _timeoutCtRegistration fields from outside may result in _tcs being incomplete or process handle leak.
+    internal sealed class ProcessRunner
+#pragma warning restore CA1001 // Types that own disposable fields should be disposable
+    {
+        private readonly IProcess _process;
+        private readonly TaskCompletionSource<string> _tcs;
+        private readonly CancellationToken _cancellationToken;
+        private readonly CancellationToken _timeoutCancellationToken;
+        private readonly CancellationTokenSource _timeoutCts;
+        private readonly CancellationTokenRegistration _ctRegistration;
+        private readonly CancellationTokenRegistration _timeoutCtRegistration;
+
+        public ProcessRunner(IProcess process, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            _process = process;
+            _cancellationToken = cancellationToken;
+            _tcs = new TaskCompletionSource<string>();
+
+            _ctRegistration = cancellationToken.Register(QueueHandleCancel);
+            if (timeout.TotalMilliseconds >= 0)
+            {
+                _timeoutCts = new CancellationTokenSource();
+                _timeoutCancellationToken = _timeoutCts.Token;
+                _timeoutCtRegistration = _timeoutCts.Token.Register(QueueHandleCancel);
+                _timeoutCts.CancelAfter(timeout);
+            }
+
+            process.Exited += (o, e) => HandleExit();
+        }
+
+        public Task<string> RunAsync()
+        {
+            if (!IsCanceled() && !_tcs.Task.IsCompleted)
+            {
+                _process.Start();
+            }
+
+            return _tcs.Task;
+        }
+
+        public string Run()
+        {
+            if (!IsCanceled() && !_tcs.Task.IsCompleted)
+            {
+                _process.Start();
+            }
+
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult().
+            return _tcs.Task.GetAwaiter().GetResult();
```

Comment:
_process.WaitForExit(()?

Reply:
`_process.WaitForExit` doesn't handle `CancellationToken` and will require separate implementation for sync and async handling of timeout.

Reply:
But you are killing the process in the cancellation callback so `WaitForExit` would return.

Reply:
Yes, `WaitForExit` will return `true` in this case, and then we will still have to figure out what has happened to the process. If your point is that we can use:
```cs
_process.WaitForExit();
return _tcs.Task.EnsureCompleted();
```
I'm not sure that we can guarantee that `_process.WaitForExit` will always return after `_tcs.Task` became completed.